src/monitoring_tests/fee_monitoring/fee_monitoring.py

```diff
@@ -0,0 +1,242 @@
+            quote_response = requests.post(
+                prod_endpoint_url,
+                headers=header,
+                json=request_dict,
+                timeout=30,
+            )
+            if quote_response.status_code != SUCCESS_CODE:
+                barn_endpoint_url = "https://barn.api.cow.fi/mainnet/api/v1/quote"
+                quote_response = requests.post(
+                    barn_endpoint_url,
+                    headers=header,
+                    json=json.dumps(request_dict),
+                    timeout=30,
+                )
+                if quote_response.status_code != SUCCESS_CODE:
+                    TemplateTest.logger.error(
+                        "Quote error: %s.", quote_response.status_code
+                    )
+        except ValueError as except_err:
+            TemplateTest.logger.error("Unhandled exception: %s.", str(except_err))
+
+        quote_json = json.loads(quote_response.text)
+
+        quote_buy_amount = int(quote_json["quote"]["buyAmount"])
+        quote_sell_amount = int(quote_json["quote"]["sellAmount"])
+        quote_fee_amount = int(quote_json["quote"]["feeAmount"])
+
+        return quote_buy_amount, quote_sell_amount, quote_fee_amount
+
+    def get_solver_solution(  # pylint: disable=unused-argument
+        self,
+        decoded_settlement: DecodedSettlement,
+        i: int,
+    ) -> Tuple[int, int, int]:
+        """
+        Given a trade, compute buy_amount, sell_amount, and fee_amount of the trade
+        as proposed by a solver resolving the instance with only that order.
+        """
+        # solver_json = json.loads()
+
+        solver_buy_amount = 0
+        solver_sell_amount = 0
+        solver_fee_amount = 0
+
+        return solver_buy_amount, solver_sell_amount, solver_fee_amount
+
+    def fee_test(self, tx_hash) -> bool:  # pylint: disable=too-many-locals
+        """
+        Given a transaction hash, check if there is a partially-fillable order in the settlement.
+        If this is the case, perform multiple tests on the execution of those orders to check if
+        the fee was set correctly.
+        TODO: add more explanations
+        """
+        # get trades via api
+        orders = TemplateTest.get_endpoint_order_data(tx_hash)
+        # loop through trades
+
+        encoded_transaction = TemplateTest.get_encoded_transaction(tx_hash)
+        decoded_settlement = TemplateTest.get_decoded_settlement_raw(
+            encoded_transaction
+        )
+
+        partially_fillable_indices = []
+        for i, order in enumerate(orders):
+            if order["partiallyFillable"]:
+                partially_fillable_indices.append(i)
+
+        if len(partially_fillable_indices) > 0:
+            # get additional data for the batch
+            receipt = TemplateTest.get_encoded_receipt(tx_hash)
+            competition_data_list = TemplateTest.get_solver_competition_data([tx_hash])
```

Comment:
What's the delay of this test? Usually competition data is only available 64 blocks after the solution was mined.

Reply:
A hash is skipped until data is returned in this call. This leads to a delay of 64 blocks. This delay is mostly a problem for the quote part of the test since the quote uses the latest state of the chain.
In princible there is no need to use the competition endpoint for the test. The reason why I am using it for now is that there seems to be no simple way at the moment to get data on fees for partially fillable orders (except for the latest fee). And there currently is a delay due to the common queue for hashes with the endpoint test.
Running the endpoint test and the fee test separately would help. Having our API return fees for different executions of partially fillable orders would solve this.

Reply:
But then we keep doing the ebbo check until the competition endpoint is available for this batch? Wouldn't this create a lot of spam in case EBBO fails?

As to how to see the fee, couldn't we just take the trade execution from the calldata or events that the settlement emits ([example](https://etherscan.io/tx/0x9e10d41f7e559e14a36b521192456e4c7c8c9b7aa3110937d28a4d43837ca36e#eventlog)) and compute the difference between the executed amounts and the amounts that should have been executed given the uniform clearing price vector? This way the main check (how economically viable was a specific settlemtn) only depends on one data source?

Reply:
But then we keep doing the ebbo check until the competition endpoint is available for this batch? Wouldn't this create a lot of spam in case EBBO fails?

As to how to see the fee, couldn't we just take the trade execution from the calldata or events that the settlement emits ([example](https://etherscan.io/tx/0x9e10d41f7e559e14a36b521192456e4c7c8c9b7aa3110937d28a4d43837ca36e#eventlog)) and compute the difference between the executed amounts and the amounts that should have been executed given the uniform clearing price vector? This way the main check (how economically viable was a specific settlemtn) only depends on one data source?

Reply:
But then we keep doing the ebbo check until the competition endpoint is available for this batch? Wouldn't this create a lot of spam in case EBBO fails?

As to how to see the fee, couldn't we just take the trade execution from the calldata or events that the settlement emits ([example](https://etherscan.io/tx/0x9e10d41f7e559e14a36b521192456e4c7c8c9b7aa3110937d28a4d43837ca36e#eventlog)) and compute the difference between the executed amounts and the amounts that should have been executed given the uniform clearing price vector? This way the main check (how economically viable was a specific settlemtn) only depends on one data source?

Reply:
But then we keep doing the ebbo check until the competition endpoint is available for this batch? Wouldn't this create a lot of spam in case EBBO fails?

As to how to see the fee, couldn't we just take the trade execution from the calldata or events that the settlement emits ([example](https://etherscan.io/tx/0x9e10d41f7e559e14a36b521192456e4c7c8c9b7aa3110937d28a4d43837ca36e#eventlog)) and compute the difference between the executed amounts and the amounts that should have been executed given the uniform clearing price vector? This way the main check (how economically viable was a specific settlemtn) only depends on one data source?

Reply:
It is true that currently we could theoretically run tests multiple times for some hashes since a hash is only removed if _both_ tests are successful. To avoid this, tests should have their own queue of hashes to check.
(At the moment the quote test for fees is the only test that -- in principle -- does not depend on the competition data. But as of right now all tests fail before checking anything if this data is missing.)

Regarding fees: If I understand correctly it is not possible to obtain the fee for limit orders from on-chain information. In some cases it is possible, e.g. in case of just one order, no internalization, and no slippage. In general, we do not store actual uniform clearing prices on chain, just the adapted prices that implicitly deduct the fees. From the latter prices, the fee cannot be inferred. (cc @harisang )

Reply:
We do encode the uniform clearing prices on chain (take https://explorer.phalcon.xyz/tx/eth/0xb1add23c49fc99f1471b61e48a1f0e6eb18f88d190144cea80dfc290ad0bcc98?line=0 for instance, it only has one partially fillable order but 4 prices, the UCP and the after fee prices for the order)

Reply:
Interesting, I did not know that. I will have a look.